### PR TITLE
Happychat: Monitor Redux actions and send whitelisted ones as customer events

### PR DESCRIPTION
**Pull request by mattwondra**
Starts sending certain Redux actions to Happychat as customer events.

There's a list of stubbed Actions in that are just `return null;`-ing — this is intentional. We will continue working on those and sending messages for the actions in separate PRs.

The actions themselves have been tested, what need code review primarily is the changes to the core middleware function (all Log and Customer events are now passing through the `sendActionLogsAndEvents` function) and the associated tests (the diff is big for this because we moved some responsibilities around in the tested functions, but some of the changes are still just copy-pasted code).

### CODE REVIEW

File: client/state/happychat/middleware.js [hunk 1]
```diff
@@ -193,6 +262,27 @@ export const sendAnalyticsLogEvent = ( connection, { getState }, { meta: { analy
 	} );
 };
 
+export const sendActionLogsAndEvents = ( connection, { getState }, action ) => {
+	const state = getState();
+
+	// If there's not an active Happychat session, do nothing
+	if ( ! isHappychatClientConnected( state ) || ! isHappychatChatAssigned( state ) ) {
+		return false;
```
Comment (southp) [inline client/state/happychat/middleware.js]: It looks like we can simply `return` here.

File: client/state/happychat/middleware.js [hunk 2]
```diff
@@ -193,6 +262,27 @@ export const sendAnalyticsLogEvent = ( connection, { getState }, { meta: { analy
...
+	}
+
+	// If there's analytics metadata attached to this action, send analytics events
+	if ( has( action, 'meta.analytics' ) ) {
+		sendAnalyticsLogEvent( connection, action );
```
Comment (southp) [inline client/state/happychat/middleware.js]: Since we don't early return here, actions with analytics data will flow through the rest of the function. Are we intentionally doing this? If not, we might want to early return here so that whoever touching analytics actions in the future can have less mental burden :)

File: client/state/happychat/selectors.js [hunk 3]
```diff
@@ -42,15 +42,15 @@ export const getHappychatTranscriptTimestamp = state => (
  * @return {String} current state value
  */
 export const getHappychatConnectionStatus = createSelector(
-	state => state.happychat.connectionStatus
+	state => get( state, 'happychat.connectionStatus' )
```
Comment (southp) [inline client/state/happychat/selectors.js]: From [the reducer side](https://github.com/Automattic/wp-calypso/blob/master/client/state/happychat/reducer.js#L153), it looks like defaulting to `'uninitialized'`. As there are more and more places reusing these state strings, ( e.g. isHappychatClientConnected` ), it might be a good timing to add those to `constants.js`, so we have a unified place to access them.

File: client/state/happychat/selectors.js [hunk 4]
```diff
@@ -42,15 +42,15 @@ export const getHappychatTranscriptTimestamp = state => (
...
 );
 
 export const isHappychatConnectionUninitialized = state => getHappychatConnectionStatus( state ) === 'uninitialized';
 
 export const isHappychatClientConnected = state => getHappychatConnectionStatus( state ) === 'connected';
 
 export const isHappychatChatAssigned = createSelector(
-	state => state.happychat.chatStatus === HAPPYCHAT_CHAT_STATUS_ASSIGNED
+	state => get( state, 'happychat.chatStatus' ) === HAPPYCHAT_CHAT_STATUS_ASSIGNED
```
Comment (southp) [inline client/state/happychat/selectors.js]: Should this default to `HAPPYCHAT_CHAT_STATUS_DEFAULT`?

File: client/state/happychat/selectors.js (hunk 3 above)
Comment (mattwondra) [inline client/state/happychat/selectors.js]: I agree that we really need to turn those into constants. Is it OK to do this in another PR though? Nothing about the logic of those strings is changing in this PR — I'm simply protecting against errors when the state tree is missing keys.

File: client/state/happychat/selectors.js (hunk 4 above)
Comment (mattwondra) [inline client/state/happychat/selectors.js]: Hm, my instinct is no — it's probably better to get `undefined` if the state tree isn't set. I worry that setting the default here, and also having it set in the reducer, means that we're defining a default value for the same state item in two places, and it would be easy to get out of sync.

What do you think?

File: client/state/happychat/middleware.js (hunk 2 above)
Comment (mattwondra) [inline client/state/happychat/middleware.js]: Yes, this is intentional. A single action may have both analytics data attached to it, and may also be an action type we want to send to the operator. For example:

```
{
    type: 'HAPPYCHAT_BLUR',  // This is the action type, which we want to send to operator
    meta: {
        analytics: [ ... ]  // This action also has analytics data attached, which we may be interested in
    }
}
```
